src/core/ruby_version.rs

```diff
@@ -181,6 +188,78 @@ macro_rules! ruby_version_v2_7_x(
     )
 );
 
+macro_rules! ruby_version_v3_0_x(
+    ($ruby_version:ident) => (
+        pub mod $ruby_version {
+            use std;
+            use bindings::$ruby_version::*;
+            use crate::core::types::ProcessMemory;
+            use failure::ResultExt;
+            use failure::Error;
+
+            get_stack_trace!(rb_execution_context_struct);
+            get_execution_context_from_vm!();
+            get_ruby_string!();
+            get_cfps!();
+            get_pos!(rb_iseq_constant_body);
+            get_lineno_2_6_0!();
+            get_stack_frame_2_5_0!();
+            stack_field_2_5_0!();
+            get_ruby_string_array_2_5_0!();
+            get_thread_id_2_5_0!();
+            get_cfunc_name!();
+
+            #[allow(non_upper_case_globals)]
+            const ruby_fl_type_RUBY_FL_USHIFT: ruby_fl_type = ruby_fl_ushift_RUBY_FL_USHIFT as i32;
+        }
+    )
+);
+
+macro_rules! get_execution_context_from_thread(
+    ($thread_type:ident) => (
+        pub fn get_execution_context<T: ProcessMemory>(ruby_current_thread_address_location: usize, _ruby_vm_address_location: usize, source: &T) -> Result<$thread_type, Error> {
+            let current_thread_addr: usize = source.copy_struct(ruby_current_thread_address_location)?;
+            let thread: $thread_type = source.copy_struct(current_thread_addr)?;
+            Ok(thread)
+        }
+    )
+);
+
+macro_rules! get_execution_context_from_vm(
+    () => (
+        pub fn get_execution_context<T: ProcessMemory>(_ruby_current_thread_address_location: usize, ruby_vm_address_location: usize, source: &T) -> Result<rb_execution_context_struct, Error> {
+            // This is a roundabout way to get the execution context address, but it helps us 
+            // avoid platform-specific structures in memory (e.g. pthread types) that would
+            // require us to maintain separate ruby-structs bindings for each platform due to
+            // their varying sizes and alignments.
+            let vm_addr: usize = source.copy_struct(ruby_vm_address_location)
+                .context(ruby_vm_address_location)?;
+            let vm: rb_vm_struct = source.copy_struct(vm_addr as usize)
+                .context(vm_addr)?;
+
+            // Seek forward in the ractor struct, looking for the main thread's address. The 
+            // execution context pointer is in the memory word just before it 
+            // (see rb_ractor_struct). The initial offset was found through experiment.
+            let initial_offset = 520;
+            for i in initial_offset..(initial_offset+500) {
+                let candidate_main_thread_address: usize = source.copy_struct(vm.ractor.main_ractor as usize + i)
```

Comment:
Would it make sense here to do 1 big copy here (into an array of addresses) instead of many small copies? It would be faster but it only happens once.

Reply:
I like that. Getting the memory addresses into a standard vector so that it can be iterated over will probably make this logic easier to follow, too.